File: project/drivers/driver/dart/lib/src/leak_detector.dart
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import 'dart:async';

class LeakDetectionConfig {
  int thresholdMs;
  bool captureStackTrace;
  int checkIntervalMs;
  void Function(String connectionId, CheckoutInfo info, int heldDurationMs)?
      onLeakDetected;

  LeakDetectionConfig({
    this.thresholdMs = 30000,
    this.captureStackTrace = false,
    this.checkIntervalMs = 10000,
    this.onLeakDetected,
  });
}

class CheckoutInfo {
  final int checkoutTime = DateTime.now().millisecondsSinceEpoch;
  final String? stackTrace;
  final Map<String, String> metadata;

  CheckoutInfo(this.metadata, {required bool captureStackTrace})
      : stackTrace = captureStackTrace ? StackTrace.current.toString() : null;

  int heldDurationMs() => DateTime.now().millisecondsSinceEpoch - checkoutTime;
}

class LeakDetectionGuard {
  final LeakDetector _detector;
  final String _connectionId;
  bool _released = false;

  LeakDetectionGuard(this._detector, this._connectionId);

  void release() {
    if (_released) return;
    _released = true;
    _detector.checkin(_connectionId);
  }
}

class LeakDetector {
  final LeakDetectionConfig config;
  final Map<String, CheckoutInfo> _checkouts = {};
  Timer? _timer;

  LeakDetector([LeakDetectionConfig? config])
      : config = config ?? LeakDetectionConfig();

  void start() {
    if (_timer != null) return;
    _timer = Timer.periodic(
        Duration(milliseconds: config.checkIntervalMs), (_) => _checkLeaks());
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  LeakDetectionGuard checkout(String connectionId,
      {Map<String, String> metadata = const {}}) {
    _checkouts[connectionId] = CheckoutInfo(Map<String, String>.from(metadata),
        captureStackTrace: config.captureStackTrace);
    return LeakDetectionGuard(this, connectionId);
  }

  void checkin(String connectionId) {
    _checkouts.remove(connectionId);
  }

  void _checkLeaks() {
    for (final entry in _checkouts.entries) {
      final heldDurationMs = entry.value.heldDurationMs();
      if (heldDurationMs > config.thresholdMs) {
        if (config.onLeakDetected != null) {
          config.onLeakDetected!(entry.key, entry.value, heldDurationMs);
          continue;
        }
        // ignore: avoid_print
        print(
            "POSSIBLE CONNECTION LEAK: conn=${entry.key} held=${heldDurationMs}ms");
      }
    }
  }
}

File: project/drivers/driver/node/src/leak_detector.ts
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

export interface LeakDetectionConfig {
  thresholdMs?: number;
  captureStackTrace?: boolean;
  checkIntervalMs?: number;
}

export class CheckoutInfo {
  readonly checkoutTime = Date.now();
  readonly stackTrace?: string;

  constructor(public readonly metadata: Record<string, unknown>, captureStackTrace: boolean) {
    if (captureStackTrace) {
      this.stackTrace = new Error().stack;
    }
  }

  heldDurationMs(): number {
    return Date.now() - this.checkoutTime;
  }
}

export class LeakDetectionGuard {
  private released = false;

  constructor(private readonly detector: LeakDetector, private readonly connectionId: string) {}

  release(): void {
    if (this.released) {
      return;
    }
    this.released = true;
    this.detector.checkin(this.connectionId);
  }
}

export class LeakDetector {
  private readonly config: Required<LeakDetectionConfig>;
  private readonly checkouts = new Map<string, CheckoutInfo>();
  private timer?: NodeJS.Timeout;

  constructor(config: LeakDetectionConfig = {}) {
    this.config = {
      thresholdMs: config.thresholdMs ?? 30_000,
      captureStackTrace: config.captureStackTrace ?? false,
      checkIntervalMs: config.checkIntervalMs ?? 10_000,
    };
  }

  start(): void {
    if (this.timer) {
      return;
    }
    this.timer = setInterval(() => this.checkLeaks(), this.config.checkIntervalMs);
    this.timer.unref?.();
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = undefined;
    }
  }

  checkout(connectionId: string, metadata: Record<string, unknown> = {}): LeakDetectionGuard {
    const info = new CheckoutInfo(metadata, this.config.captureStackTrace);
    this.checkouts.set(connectionId, info);
    return new LeakDetectionGuard(this, connectionId);
  }

  checkin(connectionId: string): void {
    this.checkouts.delete(connectionId);
  }

  activeCount(): number {
    return this.checkouts.size;
  }

  stats(): Record<string, unknown> {
    let potentialLeaks = 0;
    for (const info of this.checkouts.values()) {
      if (info.heldDurationMs() > this.config.thresholdMs) {
        potentialLeaks += 1;
      }
    }
    return { activeCheckouts: this.checkouts.size, potentialLeaks };
  }

  private checkLeaks(): void {
    for (const [connId, info] of this.checkouts.entries()) {
      if (info.heldDurationMs() > this.config.thresholdMs) {
        // Use console warning to avoid pulling in logging deps
        // eslint-disable-next-line no-console
        console.warn(`POSSIBLE CONNECTION LEAK: conn=${connId} held=${info.heldDurationMs()}ms`);
      }
    }
  }
}
